File: src/cfrpp/escfr.cpp
#include <unordered_map>

class InfoSet;
class Action;

using ISetMap = std::unordered_map<InfoSet, double>;
using ActionMap = std::unordered_map<Action, double>;

Action sample(const ActionMap &strategy);
ActionMap matching(const ActionMap &regret);

template <class Game>
double escfr(const Game &game, const int player,
             std::unordered_map<InfoSet, ActionMap> &regrets,
             std::unordered_map<InfoSet, ActionMap> &strategies) {
    if (game.terminal()) {
        return game.payoff(player);
    }

    if (game.chance()) {
        auto action = game.sample();
        return escfr(game.apply(action), player, regrets, strategies);
    }

    auto infoset = game.infoset();
    auto actions = infoset.actions();

    if (regrets.find(infoset) == regrets.end()) {
        ActionMap regret, strategy;
        for (auto action : actions) {
            regret[action] = 0;
            strategy[action] = 0;
        }

        regrets[infoset] = regret;
        strategies[infoset] = strategy;
    }

    auto R = regrets[infoset];
    auto S = strategies[infoset];
    auto strategy = matching(R);

    if (game.active() != player) {
        auto action = sample(strategy);
        auto value = escfr(game.apply(action), player, regrets, strategies);

        for (auto action : actions) {
            S[action] += strategy[action];
        }

        return value;
    }

    double value = 0;

    for (auto action : actions) {
        auto cf = escfr(game.apply(action), player, regrets, strategies);
        value += strategy[action] * cf;
        R[action] += counterfactuals[action];
    }

    for (auto action : actions) {
        R[action] -= value;
    }

    return value;
}

File: src/cfrpy/algorithms/escfr.py
from typing import cast, Generic
from typing import Dict, Callable
from dataclasses import dataclass, field
import random

from ..game import Game, Player, I, A_inv
from .cfr import matching


# Externally Sampled CFR: a version of Monte Carlo CFR
# the idea is to sample chance's and the opponent's actions, but follow all
# of the player's actions. this is also simpler than CFR.
# http://mlanctot.info/files/papers/nips09mccfr.pdf
# http://mlanctot.info/files/papers/PhD_Thesis_MarcLanctot.pdf
@dataclass
class ESCFR(Generic[A_inv, I]):
    regrets: Dict[I, Dict[A_inv, float]] = field(default_factory=dict)
    strategies: Dict[I, Dict[A_inv, float]] = field(default_factory=dict)
    touched: int = 0

    def _run_iteration(self, game: Callable[[], Game[A_inv, I]]):
        for p in (0, 1):
            self.walk(game(), cast(Player, p), self.regrets, self.strategies)

    def walk(
        self,
        game: Game[A_inv, I],
        player: Player,
        regrets: Dict[I, Dict[A_inv, float]],
        strategies: Dict[I, Dict[A_inv, float]],
    ):
        self.touched += 1

        if game.terminal:
            return game.payoff(player)

        if game.chance:
            action = game.sample()
            return self.walk(game.apply(action), player, regrets, strategies)

        infoset = game.infoset(game.active)
        actions = infoset.actions()

        if infoset not in regrets:
            regrets[infoset] = {action: 0 for action in actions}
        if infoset not in strategies:
            strategies[infoset] = {action: 0 for action in actions}

        R = regrets[infoset]
        S = strategies[infoset]

        strategy = matching(list(R.values()))

        if game.active != player:
            (action,) = random.choices(actions, weights=strategy)
            value = self.walk(game.apply(action), player, regrets, strategies)

            for action, p in zip(actions, strategy):
                S[action] += p

            return value

        cfs = {action: 0 for action in actions}
        value = 0

        for action, p in zip(actions, strategy):
            cf = self.walk(game.apply(action), player, regrets, strategies)
            cfs[action] = cf
            value += p * cf

        for action in actions:
            R[action] += cfs[action] - value

        return value
